### PR TITLE
Add EbrCell for review

### PR DESCRIPTION
**Pull request by Firstyear**
This is the implementation of an EbrCell with associated verification and memory release tests.

### CODE REVIEW

File: src/ebrcell.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    extern crate time;
+    extern crate crossbeam_utils;
+
+    use std::sync::atomic::{AtomicUsize, Ordering};
+
+    use super::EbrCell;
+    use crossbeam_utils::scoped;
+
+    #[test]
+    fn test_simple_create() {
```
Comment (Firstyear) [inline src/ebrcell.rs]: Add a test for aborting a write txn.

**Comment by Amanieu** [comment]
I personally would prefer a more "bare-bones" API where you provide the `Guard` and `Mutex` yourself:

```rust
struct EbrCell<T>(Atomic<T>);

impl<T> EbrCell {
    pub fn get<'a>(&self, guard: &'a Guard) -> &'a T {
        // just return a reference with the lifetime of the guard
    }

    pub fn set(&self, guard: &Guard, val: Box<T>) {
        // swap the new value in, and defer-free the old one
    }
}
```

We would need some way to guarantee that the `Guard` always comes from the same `Collector` though...

**Comment by Firstyear** [comment]
Hi there,

I considered this "bare-bones" style, but chose against it. There are some reasons.

Having a mutex "passed in" has really little purpose. Once we take ownership of it, no one else can use it, and there is no point "mutex" protecting the inner content, because that's already protected in the design of the Api + Guard.

Having the external Guard opens up *exactly* the issue you describe - there is no way to assert that the Guard comes from the same collector. Today, this is okay, because there is only ONE epoch collector globally, but this may change.

And this would lead to horrible memory safety issues like:

```
let guard1 = epoch_a::pin()
let guard2 = epoch_b::pin()

let a = get(guard1) ...
{
    let b = get(guard2)
}
```

Then when b goes out of scope, A could be freed *despite* loading the same pointer, because epoch_b has advanced far enough. And this is all "safe" rust as a bonus!

As a result, I decided that I didn't want to have the "consumer" think about *ANY* safety issues. It's my job to understand memory barriers and atomics and all that mess. Your life as an external consumer is to just use the api.

I wanted to make this a replacement to "RwLock<Box<T>>".

If you consider how that api would work it:

```
let x = RwLock::new(
    Box::new( DATA )
);

{
    let read_guard = x.read_lock()
    //use the content of read_guard
}
```

Similar story for write. I wanted to emulate the ease of use of the RwLock api here, and that's why I designed the external interface to be a little be like this. You declare your intent to read or write, and the safety happens *for you*, you don't think about any of it.

In summary - the bare-bones api adds overhead to the programmer's ergonomic use, it adds potential for unsafety, and it has zero appreciable benefit over the design I chose today.

Thanks,

William

**Comment by Firstyear** [comment]
@Vtec234 Thanks for your review recently mate. I finally found some time and applied your suggestions. Really appreciate it!
